src/data/fintype/fiber.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Johan Commelin, Scott Morrison
+-/
+import data.fintype.basic
+
+/-!
+# The fibre over a point, when the domain is a fintype.
+
+We define the fiber of a function `f : α → β` over a point `b : β`, when we have `fintype α`,
+as a `finset α`, and prove some basic properties.
+-/
+
+open function
+
+namespace fintype
+/--
+The fiber of function `f : α → β` over a point `b : β` is a `finset α` when we have `fintype α`.
+-/
+def fiber {α : Type*} [fintype α] {β : Type*} [decidable_eq β] (f : α → β) (b : β) : finset α :=
+set.to_finset { a | f a = b }
+
+@[simp]
+lemma mem_fiber {α : Type*} [fintype α] {β : Type*} [decidable_eq β] (f : α → β) (b : β) (a : α) :
+  a ∈ fintype.fiber f b ↔ f a = b :=
+by simp [fiber]
+
+@[simp]
+lemma card_fiber {α : Type*} [fintype α] {β : Type*} [decidable_eq β] (f : α → β) (b : β) :
+  finset.card (fiber f b) = fintype.card { a | f a = b } :=
+set.to_finset_card { a | f a = b }
+
+lemma fibers_disjoint {α : Type*} [fintype α] [decidable_eq α] {β : Type*} [decidable_eq β] (f : α → β) (b₁ b₂ : β) (h : b₁ ≠ b₂) :
+  disjoint (fiber f b₁) (fiber f b₂) :=
+begin
+  dsimp [disjoint],
+  rintros a w,
+  simp [fiber] at w,
+  exact false.elim (h (w.1.symm.trans w.2))
+end
+
+@[simp]
+lemma fiber_prod_fst {α : Type*} [fintype α] [decidable_eq α] {β : Type*} [fintype β] (a : α) :
```

Comment:
I don't think `\a` has to be a fintype here, right?
I also just noticed that `finset.product` exists, so we can relate this to `(finset.singleton a).product (finset.univ : finset \b)`.

Finally, I think it would be good to have a simp-lemma about the cardinality of these fibers.

Reply:
The source of `prod.fst` won't be a fintype if \a isn't.